extension/extesionhelper/factory_test.go

```diff
@@ -0,0 +1,78 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensionhelper
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+)
+
+const typeStr = "test"
+
+var (
+	defaultCfg = &configmodels.ExtensionSettings{
+		TypeVal: typeStr,
+		NameVal: typeStr,
+	}
+	nopTracesExporter = exportertest.NewNopTraceExporter()
+)
+
+func TestNewFactory(t *testing.T) {
+	factory := NewFactory(
+		typeStr,
+		defaultConfig,
+		createExtension)
+	assert.EqualValues(t, typeStr, factory.Type())
+	assert.EqualValues(t, defaultCfg, factory.CreateDefaultConfig())
+	_, err := factory.CreateExtension(context.Background(), component.ExtensionCreateParams{}, defaultCfg)
+	assert.NoError(t, err)
+}
+
+func TestNewFactory_WithConstructors(t *testing.T) {
+	factory := NewFactory(
+		typeStr,
+		defaultConfig,
+		createExtension,
+		WithCustomUnmarshaler(customUnmarshaler))
+	assert.EqualValues(t, typeStr, factory.Type())
+	assert.EqualValues(t, defaultCfg, factory.CreateDefaultConfig())
+
+	fu, ok := factory.(component.ConfigUnmarshaler)
+	assert.True(t, ok)
+	assert.Equal(t, errors.New("my error"), fu.Unmarshal(nil, nil))
+
+	_, err := factory.CreateExtension(context.Background(), component.ExtensionCreateParams{}, defaultCfg)
+	assert.NoError(t, err)
+}
+
+func defaultConfig() configmodels.Extension {
+	return defaultCfg
+}
+
+func createExtension(context.Context, component.ExtensionCreateParams, configmodels.Extension) (component.ServiceExtension, error) {
+	return nopTracesExporter, nil
```

Comment:
It doesn't really matter for this test I guess but it seems a bit weird that `nopTracesExporter` implements `component.ServiceExtension` - is there a better mock object that can be used?

Reply:
Also would be good to validate the object returned here is the one returned by `factory.CreateExtension`

Reply:
Done.